### PR TITLE
CY-821 Allow resuming failed/cancelled executions

### PR DESCRIPTION
**Pull request by tehasdf**
Resuming a failed/cancelled executions (needs to be done manually,
and is called a "force resume" as opposed to a regular automatic
"resume" which happens on mgmtworker restarts) resets the failed
operations to pending and sets their restart count to 0.

### CODE REVIEW

File: rest-service/manager_rest/test/endpoints/test_executions.py [hunk 1]
```diff
@@ -767,3 +770,69 @@ def test_install_empty_blueprint(self):
...
+
+        operations = self.sm.list(models.Operation, get_all_results=True)
+        self.assertEqual(len(operations), 1)
+        operation = operations[0]
+        self.assertEqual(operation.state, cloudify_tasks.TASK_PENDING)
+        self.assertEqual(operation.parameters['current_retries'], 0)
+
+        execution = self.sm.get(models.Execution, execution.id)
+        self.assertEqual(execution.status, ExecutionState.PENDING)
+
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_executions.py]: what about tests that are trying to resume execution that you could not (pending and etc)?

File: rest-service/manager_rest/resource_manager.py [hunk 2]
```diff
@@ -466,10 +467,42 @@ def delete_deployment(self,
...
+
+        All operations that were failed are going to be retried,
+        the execution itself is going to be set to pending again.
+
+        :return: Whether to continue with running the execution
+        """
+        if execution.status in (ExecutionState.FAILED,
+                                ExecutionState.CANCELLED):
+            execution.status = ExecutionState.PENDING
+            self.sm.update(execution, modified_attrs=('status',))
```
[the remark this answers is not in the document]
Comment (tehasdf) [reply rest-service/manager_rest/resource_manager.py]: it actually is, though! it needs to be an iterable of attribute names, and `('status')` without the comma would be just the string "status" (not a 1-tuple)

[the remark this answers is not in the document]
Comment (tyacbovi) [reply rest-service/manager_rest/resource_manager.py]: Interesting

File: rest-service/manager_rest/test/endpoints/test_executions.py (hunk 1 above)
Comment (tehasdf) [reply rest-service/manager_rest/test/endpoints/test_executions.py]: added some

File: rest-service/manager_rest/resource_manager.py [hunk 3]
```diff
@@ -466,16 +467,46 @@ def delete_deployment(self,
...
+                    operation.parameters['current_retries'] = 0
+                    self.sm.update(operation,
+                                   modified_attrs=('parameters', 'state'))
+
+    def resume_execution(self, execution_id, force=False):
+        execution = self.sm.get(models.Execution, execution_id)
+        if force:
+            if execution.status not in (ExecutionState.CANCELLED,
+                                        ExecutionState.FAILED):
+                raise manager_exceptions.ConflictError(
```
Comment (tyacbovi) [inline rest-service/manager_rest/resource_manager.py]: Maybe add more details on the execution

File: rest-service/manager_rest/test/endpoints/test_executions.py [hunk 4]
```diff
@@ -767,3 +770,128 @@ def test_install_empty_blueprint(self):
...
+        self.assertEqual(operation.parameters['current_retries'], 0)
+
+        execution = self.sm.get(models.Execution, execution.id)
+        # started or terminated, it might have already finished by the time
+        # the test was done
+        self.assertIn(execution.status,
+                      (ExecutionState.STARTED, ExecutionState.TERMINATED))
+
+    def test_resume_force_failed(self):
+        """Force-resume resets operation state and restart count"""
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_executions.py]: there is repetitiveness in the tests, but I don't like to write more framework for few tests

**Pull request merged**